File: lib/std/js/io.js
import * as ops from 'std:io.go'

const DEFAULT_CHUNK_SIZE = 64 * 1024
const textEncoder = typeof TextEncoder === 'function' ? new TextEncoder() : null

const readerState = new WeakMap()
const writerState = new WeakMap()

function normalizeHandle(handle, registerFn, type) {
    if (typeof handle !== 'string') {
        throw new TypeError(`${type} handle must be a string`)
    }
    return registerFn(handle)
}

function toUint8Array(value) {
    if (value == null) {
        throw new TypeError('chunk is required')
    }

    if (value instanceof Uint8Array) {
        return value
    }

    if (value instanceof ArrayBuffer) {
        return new Uint8Array(value)
    }

    if (ArrayBuffer.isView(value)) {
        return new Uint8Array(value.buffer, value.byteOffset, value.byteLength)
    }

    if (typeof value === 'string') {
        if (textEncoder) {
            return textEncoder.encode(value)
        }

        const buffer = new Uint8Array(value.length)
        for (let i = 0; i < value.length; i += 1) {
            buffer[i] = value.charCodeAt(i) & 0xff
        }
        return buffer
    }

    throw new TypeError(`Unsupported chunk type: ${typeof value}`)
}

function isHandleNotFound(error) {
    if (error == null) {
        return false
    }

    const message = typeof error === 'string' ? error : error.message
    return typeof message === 'string' && message.includes('handle not found')
}

async function closeReaderState(state, force = false) {
    if (state.closed) {
        return
    }

    state.closed = true
    if (state.finalizerToken) {
        state.finalizerToken = null
    }

    if (state.autoClose || force) {
        try {
            ops.op_closeHandle(state.handle)
        } catch (error) {
            if (!isHandleNotFound(error)) {
                throw error
            }
        }
    }
}

async function closeWriterState(state, force = false) {
    if (state.closed) {
        return
    }

    state.closed = true
    if (state.finalizerToken) {
        state.finalizerToken = null
    }

    if (state.autoClose || force) {
        try {
            ops.op_closeHandle(state.handle)
        } catch (error) {
            if (!isHandleNotFound(error)) {
                throw error
            }
        }
    }
}

class Reader {
    highWaterMark = 64 * 1024
    constructor(handle, options = {}) {
        const { autoClose = true, highWaterMark = 64 * 1024 } = options;
        this.highWaterMark = highWaterMark;
        const normalized = normalizeHandle(handle, ops.op_registerReader, 'Reader')

        const state = {
            handle: normalized,
            closed: false,
            autoClose,
            finalizerToken: null,
        }

        if (autoClose) {
            state.finalizerToken = Como.finalizer(() => {
                ops.op_closeHandle(handle)
            })
        }

        readerState.set(this, state)
    }

    static fromHandle(handle, options) {
        return new Reader(handle, options)
    }

    static fromStdin(options) {
        const handle = ops.op_readerFromStdin()
        return new Reader(handle, options)
    }

    static fromFile(path, options) {
        const handle = ops.op_readerFromFile(path)
        return new Reader(handle, options)
    }

    static fromCallback(callback, options) {
        if (typeof callback !== 'function') {
            throw new TypeError('callback must be a function')
        }

        const handle = ops.op_readerFromCallback(callback)
        return new Reader(handle, options)
    }

    get closed() {
        const state = readerState.get(this)
        return !!(state && state.closed)
    }

    get handle() {
        const state = readerState.get(this)
        return state && state.handle
    }

    async read(size) {
        size = size ?? this.highWaterMark;
        const state = readerState.get(this)
        if (!state) {
            throw new Error('Reader state missing')
        }

        if (state.closed) {
            return null
        }

        const result = await ops.op_readerRead(state.handle, size);
        let chunk = result.chunk
        const { done } = result

        if (chunk != null) {
            if (chunk instanceof Uint8Array) {
                // already fine
            } else if (chunk instanceof ArrayBuffer) {
                chunk = new Uint8Array(chunk)
            } else if (ArrayBuffer.isView(chunk)) {
                chunk = new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength)
            } else {
                throw new TypeError(`Unexpected reader chunk type: ${typeof chunk}`)
            }
        }

        if (done) {
            await closeReaderState(state)
        }

        if (chunk == null || chunk.length === 0) {
            return done ? null : new Uint8Array()
        }

        return chunk
    }

    async pipeTo(writer, options = {}) {
        const state = readerState.get(this)
        if (!state || state.closed) {
            throw new Error('reader is closed')
        }

        if (!(writer instanceof Writer)) {
            throw new TypeError('writer must be an instance of Writer')
        }

        const writerInfo = writerState.get(writer)
        if (!writerInfo || writerInfo.closed) {
            throw new Error('writer is closed')
        }

        const args = [state.handle, writerInfo.handle]
        if (options && options.chunkSize != null) {
            const chunkSize = Number(options.chunkSize)
            if (!Number.isFinite(chunkSize) || chunkSize <= 0) {
                throw new TypeError('chunkSize must be a positive number')
            }
            args.push(chunkSize)
        }

        return ops.op_readerPipe(...args)
    }

    async close() {
        const state = readerState.get(this)
        if (!state) {
            return
        }

        await closeReaderState(state, true)
    }

    [Symbol.asyncIterator]() {
        return {
            reader: this,
            async next() {
                const chunk = await this.reader.read()
                if (chunk === null) {
                    return { value: undefined, done: true }
                }
                return { value: chunk, done: false }
            },
        }
    }

    toReadableStream(streamOptions) {
        if (typeof ReadableStream !== 'function') {
            throw new Error('ReadableStream is not available in this environment')
        }

        const reader = this
        return new ReadableStream(
            {
                async pull(controller) {
                    const chunk = await reader.read()
                    if (chunk === null) {
                        controller.close()
                        return
                    }
                    controller.enqueue(chunk)
                },
                async cancel() {
                    await reader.close()
                },
            },
            streamOptions,
        )
    }
}

class Writer {
    constructor(handle, options = {}) {
        const { autoClose = true } = options
        const normalized = normalizeHandle(handle, ops.op_registerWriter, 'Writer')

        const state = {
            handle: normalized,
            closed: false,
            autoClose,
            finalizerToken: null,
        }

        if (autoClose) {
            state.finalizerToken = Como.finalizer(() => {
                ops.op_closeHandle(handle)
            })

            // writerFinalizer.register(this, normalized, state.finalizerToken)
        }

        writerState.set(this, state)
    }

    static fromHandle(handle, options) {
        return new Writer(handle, options)
    }

    static fromCallback(callback, options) {
        if (typeof callback !== 'function') {
            throw new TypeError('callback must be a function')
        }

        const handle = ops.op_writerFromCallback((chunk) => {
            return callback(toUint8Array(chunk))
        })

        return new Writer(handle, options)
    }

    static fromFile(path, options) {
        const handle = ops.op_writerFromFile(path)
        return new Writer(handle, options)
    }

    get closed() {
        const state = writerState.get(this)
        return !!(state && state.closed)
    }

    get handle() {
        const state = writerState.get(this)
        return state && state.handle
    }

    async write(chunk) {
        const state = writerState.get(this)
        if (!state) {
            throw new Error('Writer state missing')
        }

        if (state.closed) {
            throw new Error('writer is closed')
        }

        const payload = toUint8Array(chunk)
        return ops.op_writerWrite(state.handle, payload)
    }

    async close() {
        const state = writerState.get(this)
        if (!state) {
            return
        }

        await closeWriterState(state, true)
    }

    toWritableStream(streamOptions) {
        if (typeof WritableStream !== 'function') {
            throw new Error('WritableStream is not available in this environment')
        }

        const writer = this
        return new WritableStream(
            {
                async write(chunk) {
                    await writer.write(chunk)
                },
                async close() {
                    await writer.close()
                },
                async abort(reason) {
                    await writer.close()
                    return reason
                },
            },
            streamOptions,
        )
    }
}

export function fromReaderHandle(handle, options) {
    return Reader.fromHandle(handle, options)
}

export function fromWriterHandle(handle, options) {
    return Writer.fromHandle(handle, options)
}

export function readerFromCallback(callback, options) {
    return Reader.fromCallback(callback, options)
}

export function writerFromCallback(callback, options) {
    return Writer.fromCallback(callback, options)
}

export function getReaderHandle(reader) {
    if (reader instanceof Reader) {
        return reader.handle
    }
    if (typeof reader === 'string') {
        return ops.op_registerReader(reader)
    }
    throw new TypeError('expected Reader instance or handle string')
}

export function getWriterHandle(writer) {
    if (writer instanceof Writer) {
        return writer.handle
    }
    if (typeof writer === 'string') {
        return ops.op_registerWriter(writer)
    }
    throw new TypeError('expected Writer instance or handle string')
}

export { Reader, Writer, DEFAULT_CHUNK_SIZE }

export default {
    Reader,
    Writer,
    DEFAULT_CHUNK_SIZE,
    fromReaderHandle,
    fromWriterHandle,
    readerFromCallback,
    writerFromCallback,
    getReaderHandle,
    getWriterHandle,
}

File: lib/std/io.go
package std

import (
	"errors"
	"io"
	"os"
	"strings"

	"github.com/comoland/como/js"
	iohandle "github.com/comoland/como/lib/internal"
)

func registerIo(ctx *js.Context) {
	mod := ctx.NewModule("std:io.go")

	exports := map[string]any{}

	exports["op_registerReader"] = func(args js.Arguments) interface{} {
		handleID := args.GetString(0)
		if handleID == "" {
			return ctx.Throw("reader handle id must be a non-empty string")
		}

		_, err := iohandle.GetReader(handleID)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return handleID
	}

	exports["op_registerWriter"] = func(args js.Arguments) interface{} {
		handleID := args.GetString(0)
		if handleID == "" {
			return ctx.Throw("writer handle id must be a non-empty string")
		}

		_, err := iohandle.GetWriter(handleID)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return handleID
	}

	exports["op_readerFromCallback"] = func(args js.Arguments) interface{} {
		cb := ctx.Writer(args.GetValue(0))
		if cb == nil {
			return ctx.Throw("callback must be a function")
		}

		// reader := iohandle.NewCallbackReader(ctx, cb)
		// return iohandle.NewHandle(reader, nil, reader)
		return iohandle.NewHandle(cb, nil, cb)
	}

	exports["op_writerFromCallback"] = func(args js.Arguments) interface{} {
		cb := ctx.Writer(args.GetValue(0))
		if cb == nil {
			return ctx.Throw("callback must be a function")
		}

		// writer := iohandle.NewCallbackWriter(ctx, cb)
		return iohandle.NewHandle(nil, cb, cb)
	}

	exports["op_writerFromFile"] = func(args js.Arguments) interface{} {
		path := args.GetString(0)
		file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return iohandle.NewWriterHandle(file)
	}

	exports["op_readerFromStdin"] = func(args js.Arguments) interface{} {
		return iohandle.NewReaderHandle(os.Stdin)
	}

	exports["op_readerFromString"] = func(args js.Arguments) interface{} {
		str := args.GetString(0)
		return iohandle.NewReaderHandle(strings.NewReader(str))
	}

	exports["op_readerFromFile"] = func(args js.Arguments) interface{} {
		path := args.GetString(0)
		file, err := os.Open(path)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		return iohandle.NewReaderHandle(file)
	}

	exports["op_readerRead"] = func(args js.Arguments) interface{} {
		handleID := args.GetString(0)
		if handleID == "" {
			return ctx.Throw("reader handle id must be a non-empty string")
		}

		size := iohandle.DefaultChunkSize
		if args.Len() > 1 {
			switch v := args.Get(1).(type) {
			case int64:
				if v > 0 {
					size = int(v)
				}
			case float64:
				if v > 0 {
					size = int(v)
				}
			}
		}

		return ctx.Async(func(async js.Promise) {
			reader, err := iohandle.GetReader(handleID)
			if err != nil {
				async.Reject(err.Error())
				return
			}

			buf := make([]byte, size)
			n, readErr := reader.Read(buf)
			if n > 0 {
				buf = buf[:n]
			} else {
				buf = nil
			}

			if readErr != nil && !errors.Is(readErr, io.EOF) {
				async.Reject(readErr.Error())
				return
			}

			result := map[string]any{
				"chunk":     buf,
				"bytesRead": n,
				"done":      errors.Is(readErr, io.EOF),
			}

			async.Resolve(result)
		})
	}

	exports["op_writerWrite"] = func(args js.Arguments) interface{} {
		handleID := args.GetString(0)
		if handleID == "" {
			return ctx.Throw("writer handle id must be a non-empty string")
		}

		data, err := args.GetTypedArray(1)
		if err != nil {
			return ctx.Throw(err.Error())
		}

		payload := make([]byte, len(data))
		copy(payload, data)

		return ctx.Async(func(async js.Promise) {
			writer, getErr := iohandle.GetWriter(handleID)
			if getErr != nil {
				async.Reject(getErr.Error())
				return
			}

			n, writeErr := writer.Write(payload)
			if writeErr != nil {
				async.Reject(writeErr.Error())
				return
			}

			async.Resolve(n)
		})
	}

	exports["op_readerPipe"] = func(args js.Arguments) interface{} {
		srcID := args.GetString(0)
		dstID := args.GetString(1)

		chunkSize := iohandle.DefaultChunkSize
		if args.Len() > 2 {
			switch v := args.Get(2).(type) {
			case int64:
				if v > 0 {
					chunkSize = int(v)
				}
			case float64:
				if v > 0 {
					chunkSize = int(v)
				}
			}
		}

		return ctx.Async(func(async js.Promise) {
			reader, err := iohandle.GetReader(srcID)
			if err != nil {
				async.Reject(err.Error())
				return
			}

			writer, err := iohandle.GetWriter(dstID)
			if err != nil {
				async.Reject(err.Error())
				return
			}

			var buf []byte
			if chunkSize > 0 {
				buf = make([]byte, chunkSize)
			}

			n, copyErr := io.CopyBuffer(writer, reader, buf)
			if copyErr != nil && !errors.Is(copyErr, io.EOF) {
				async.Reject(copyErr.Error())
				return
			}

			async.Resolve(map[string]any{
				"bytesCopied": n,
				"done":        errors.Is(copyErr, io.EOF),
			})
		})
	}

	exports["op_closeHandle"] = func(args js.Arguments) interface{} {
		handleID := args.GetString(0)
		if handleID == "" {
			return ctx.Throw("handle id must be a non-empty string")
		}

		if err := iohandle.ReleaseIOHandle(handleID); err != nil {
			return ctx.Throw(err.Error())
		}

		return true
	}

	mod.Exports(exports)
	mod.Export("default", exports)
}
